### PR TITLE
Model submission for explorer_r3_sensor_config_1 from Explorer

### PR DESCRIPTION
**Pull request by HongbiaoZ**
Make some changes according to the issues of the old pull request.

### CODE REVIEW

File: submitted_models/explorer_r3_sensor_config_1/model.sdf [hunk 1]
```diff
@@ -226,7 +242,7 @@
               </ray>
       </sensor>
       <velocity_decay/>
-      <sensor name="rs_back" type="rgbd_camera">
+      <sensor name="rs_rear" type="rgbd_camera">
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/model.sdf:245]: This line should be reverted to `rs_back` or the instances of `rs_back` in `vehicle_topics.launch` will need to be updated to `rs_rear`.

**Review by acschang — CHANGES_REQUESTED** [review]
Thank you for revising your model according to the issues of the other PR.  The following issues are still outstanding:

- Please add the [wheel slip plugin](https://github.com/osrf/subt/pull/569); if you need assistance with this we can add it as part of the later review process.
- Please resolve all instances of floating sensors and sensor mounts by affixing to the robot.
- Please include a 2nd sensor configuration with breadcrumbs as indicated in the `specifications.md` file.
- Please remove the links and statements related to the private repository referenced in the `README.md` for the vehicle configuration and odometry, permit public access to the linked repository, or include the associated files in this submission.
- Please resolve repeated incorrect names for RGBD publishers and the missing (omitted) input topic remaps in the `vehicle_topics.launch` file.  Please resolve conflicting naming of `rs_back` and `rs_rear` between the `model.sdf` and `vehicle_topics.launch` file that is preventing the rear RGBD sensor from publishing data.
- If this robot has similar visual elements to the R2 robot, please use meshes and/or materials from the R2 model to better match the robot's physical appearance. Meshes can be re-scaled as necessary.

File: submitted_models/explorer_r3_sensor_config_1/model.sdf [hunk 2]
```diff
@@ -61,6 +61,22 @@
...
+                  <metalness_map>materials/textures/R2_Metalness.png</metalness_map>
+                  <roughness_map>materials/textures/R2_Roughness.png</roughness_map>
+                </metal>
+            </pbr>
+            <script>
+                <uri>materials/scripts/</uri>
+                <uri>materials/textures/</uri>
+                <name>UrbanTile/R2_Diffuse</name>
+            </script>
+        </material-->
```
[the remark this answers is not in the document]
Comment (HongbiaoZ) [reply submitted_models/explorer_r3_sensor_config_1/model.sdf:79]: I tried to use the same material as R2 while it didn't work. Thus I commented them. Could you help to figure out the reason?

File: submitted_models/explorer_r3_sensor_config_1/model.sdf (hunk 1 above)
Comment (HongbiaoZ) [reply submitted_models/explorer_r3_sensor_config_1/model.sdf:245]: I have changed all rs_back to rs_rear now.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 3]
```diff
@@ -93,24 +86,20 @@
...
-        <remap from="input/camera_info" to="color/camera_info" />
         <remap from="output/camera_info" to="optical/color/camera_info" />
       </node>
       <node
         pkg="subt_ros"
         type="optical_frame_publisher"
         name="optical_frame_publisher_rgbd_depth_left">
-        <remap from="input/image" to="depth/image" />
         <remap from="output/image" to="optical/depth/image" />
-        <remap from="input/camera_info" to="depth/camera_info" />
```
[the remark this answers is not in the document]
Comment (HongbiaoZ) [reply submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch]: I get them back.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 4]
```diff
@@ -93,24 +86,20 @@
         pkg="subt_ros"
         type="optical_frame_publisher"
         name="optical_frame_publisher_rgbd_image_left">
```
[the remark this answers is not in the document]
Comment (HongbiaoZ) [reply submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch]: I have changed them all in the latest commit.

[the remark this answers is not in the document]
Comment (HongbiaoZ) [reply submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch]: For the remained issues, I fixed part of them. While there are several issues that I'm not sure how to fix them. For the first one, I need your help to add this plugin. For the second one, I don't understand what you mean "floating sensors". The sensors are all fixed to the base_link. For the third issue, I don't know how to add breadcrumbs to the configuration. Could you help to create a new configuration as you do for R2?  For the last one, I tried to use the materials of R2 in the model.sdf, for example line 64 to line 79, while it didn't work. So I comment them and change back to the old one.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 5]
```diff
@@ -85,15 +85,17 @@
       <node
         pkg="subt_ros"
         type="optical_frame_publisher"
-        name="optical_frame_publisher_rgbd_image_left">
+        name="optical_frame_publisher_rgbd_image_front">
         <remap from="output/image" to="optical/color/image" />
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch:89]: The optical frame publisher needs remaps as well, for example:
```
<remap from="input/image" to="color/image" />
<remap from="input/camera_info" to="color/camera_info" />
```

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 6]
```diff
@@ -223,15 +229,17 @@
       <node
         pkg="subt_ros"
         type="optical_frame_publisher"
-        name="optical_frame_publisher_rgbd_image_left">
+        name="optical_frame_publisher_rgbd_image_right">
         <remap from="output/image" to="optical/color/image" />
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch:233]: The optical frame publisher needs remaps as well, for example:
```
<remap from="input/image" to="color/image" />
<remap from="input/camera_info" to="color/camera_info" />
```

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 7]
```diff
@@ -150,42 +154,44 @@
...
         args="$(arg sensor_prefix)/rs_back/depth_image">
-        <remap from="$(arg sensor_prefix)/rs_back/depth_image" to="depth/image"/>
+        <remap from="$(arg sensor_prefix)/rs_rear/depth_image" to="depth/image"/>
       </node>
       <node
         pkg="subt_ros"
         type="optical_frame_publisher"
-        name="optical_frame_publisher_rgbd_image_left">
+        name="optical_frame_publisher_rgbd_image_rear">
         <remap from="output/image" to="optical/color/image" />
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch:185]: The optical frame publisher needs remaps as well, for example:
```
<remap from="input/image" to="color/image" />
<remap from="input/camera_info" to="color/camera_info" />
```

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        type="image_bridge"
+        name="ros_ign_image_depth"
+        args="$(arg sensor_prefix)/rs_left/depth_image">
+        <remap from="$(arg sensor_prefix)/rs_left/depth_image" to="depth/image"/>
+      </node>
+      <node
+        pkg="subt_ros"
+        type="optical_frame_publisher"
+        name="optical_frame_publisher_rgbd_image_left">
+        <remap from="output/image" to="optical/color/image" />
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch:137]: The optical frame publisher needs remaps as well, for example:
```
<remap from="input/image" to="color/image" />
<remap from="input/camera_info" to="color/camera_info" />
```

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch (hunk 4 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch]: Here are a couple of examples of floating sensors on the Explorer R3 model:
explorer_r3_floating_sensors_1
explorer_r3_floating_sensors_2

Both the LIDAR and the LIDAR mount are not affixed to the vehicle as shown in the above screenshots.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        type="parameter_bridge"
+        name="ros_ign_bridge_camera"
+        args="$(arg sensor_prefix)/rs_rear/camera_info@sensor_msgs/CameraInfo[ignition.msgs.CameraInfo">
+        <remap from="$(arg sensor_prefix)/rs_rear/camera_info" to="color/camera_info"/>
+      </node>
+      <node
+        pkg="ros_ign_image"
+        type="image_bridge"
+        name="ros_ign_image_camera"
+        args="$(arg sensor_prefix)/rs_back/image">
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch:164]: This instance of `rs_back` needs to be changed to `rs_rear`.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+        <remap from="$(arg sensor_prefix)/rs_rear/image" to="color/image"/>
+      </node>
+      <node
+        pkg="ros_ign_bridge"
+        type="parameter_bridge"
+        name="ros_ign_bridge_rgbd"
+        args="$(arg sensor_prefix)/rs_back/points@sensor_msgs/PointCloud2[ignition.msgs.PointCloudPacked">
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch:171]: This instance of `rs_back` needs to be changed to `rs_rear`.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+        <remap from="$(arg sensor_prefix)/rs_rear/points" to="points"/>
+      </node>
+      <node
+        pkg="ros_ign_image"
+        type="image_bridge"
+        name="ros_ign_image_depth"
+        args="$(arg sensor_prefix)/rs_back/depth_image">
```
Comment (acschang) [inline submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch:178]: This instance of `rs_back` needs to be changed to `rs_rear`.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch (hunk 3 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch]: Please check the notes below for cases where they were not re-added for the color cameras while they are needed.

**Review by acschang — CHANGES_REQUESTED** [review]
Please adjust the remaining instances of `rs_back` to `rs_rear`.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch (hunk 4 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch]: Thank you for starting on the requested changes.  We will help with the outstanding issues.

- A starting point for adding the wheel slip plugin is to reference the [Explorer R2's wheel slip plugin](https://github.com/osrf/subt/blob/master/submitted_models/explorer_r2_sensor_config_1/launch/spawner.rb#L61) and then tune the parameters for the Explorer R3 model.
- An example of the addition of breadcrumbs for the second R2 configuration can be found [here](https://github.com/osrf/subt/tree/master/submitted_models/explorer_r2_sensor_config_2).  Adding a second configuration with breadcrumbs is merely an additional catkin package with a modified `spawner.rb` from the first configuration to include the breadcrumb plugin.

I'll dig and/or ask for some additional assistance to look into the issues you're experiencing with the mesh.

**Comment by acschang** [comment]
I'm pulling your response and my reply to the conversation out of the resolved thread for tracking changes.

> For the remained issues, I fixed part of them. While there are several issues that I'm not sure how to fix. For the first one, I need your help to add this plugin. For the second one, I don't understand what you mean "floating sensors". The sensors are all fixed to the base_link. For the third issue, I don't know how to add breadcrumbs to the configuration. Could you help to create a new configuration as you do for R2? For the last one, I tried to use the materials of R2 in the model.sdf, for example line 64 to line 79, while it didn't work. So I comment them and change back to the old one.

Thank you for starting on the requested changes.  We will help with the outstanding issues.

- Both the LIDAR and the LIDAR mount are not affixed to the vehicle. Here are a couple of examples of floating sensors on the Explorer R3 model:
explorer_r3_floating_sensors_1
explorer_r3_floating_sensors_2
- A starting point for adding the wheel slip plugin is to reference the [Explorer R2's wheel slip plugin](https://github.com/osrf/subt/blob/master/submitted_models/explorer_r2_sensor_config_1/launch/spawner.rb#L61) and then tune the parameters for the Explorer R3 model.
- An example of the addition of breadcrumbs for the second R2 configuration can be found [here](https://github.com/osrf/subt/tree/master/submitted_models/explorer_r2_sensor_config_2).  Adding a second configuration with breadcrumbs is merely an additional catkin package with a modified `spawner.rb` from the first configuration to include the breadcrumb plugin.

I'll dig and/or ask for some additional assistance to look into the issues you're experiencing with the mesh.

File: submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch (hunk 4 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/explorer_r3_sensor_config_1/launch/vehicle_topics.launch]: The issue in this thread has been resolved so I'm closing this thread.  Let's talk about the outstanding changes in the thread for the entire PR.

**Comment by acschang** [comment]
Continued in #784

**Pull request closed without merging**